File: RepoTransBench/source_projects/Python/picklepete_pyicloud/public_tests/test_public_utils.py
import pytest
import sys

import pyicloud.utils as utils
from pyicloud.exceptions import PyiCloudNoStoredPasswordAvailableException

class PublicDummyKeyring:
    def __init__(self):
        self.saved = {}
        self.deleted = []
    def get_password(self, system, username):
        return self.saved.get(username, None)
    def set_password(self, system, username, password):
        self.saved[username] = password
        return "store"
    def delete_password(self, system, username):
        self.deleted.append(username)
        return "removed"

@pytest.fixture(autouse=True)
def patch_keyring(monkeypatch):
    dkr = PublicDummyKeyring()
    monkeypatch.setattr(utils, "keyring", dkr)
    return dkr

def test_get_password_from_keyring_success(patch_keyring):
    patch_keyring.saved["baz"] = "qux"
    assert utils.get_password_from_keyring("baz") == "qux"

def test_get_password_from_keyring_failure(patch_keyring):
    with pytest.raises(PyiCloudNoStoredPasswordAvailableException):
        utils.get_password_from_keyring("does-not-exist")

def test_password_exists_in_keyring_true(patch_keyring):
    patch_keyring.saved["public_user"] = "public_pw"
    assert utils.password_exists_in_keyring("public_user") is True

def test_password_exists_in_keyring_false(patch_keyring):
    assert utils.password_exists_in_keyring("anonymous") is False

def test_store_password_in_keyring(patch_keyring):
    out = utils.store_password_in_keyring("user2", "passwd2")
    assert patch_keyring.saved["user2"] == "passwd2"
    assert out == "store"

def test_delete_password_in_keyring(patch_keyring):
    patch_keyring.saved["deleteme"] = "secret"
    out = utils.delete_password_in_keyring("deleteme")
    assert "deleteme" in patch_keyring.deleted
    assert out == "removed"

def test_underscore_to_camelcase_basic():
    assert utils.underscore_to_camelcase("foo_bar_baz") == "fooBarBaz"
    assert utils.underscore_to_camelcase("Bar_c", initial_capital=True) == "BarC"

def test_get_password_interactive_false(monkeypatch):
    # disables interactive mode, keyring throws
    monkeypatch.setattr(utils, "get_password_from_keyring", lambda _u: (_ for _ in ()).throw(PyiCloudNoStoredPasswordAvailableException()))
    with pytest.raises(PyiCloudNoStoredPasswordAvailableException):
        utils.get_password("userx", interactive=False)

def test_get_password_interactive_true(monkeypatch):
    monkeypatch.setattr(utils, "get_password_from_keyring", lambda _u: (_ for _ in ()).throw(PyiCloudNoStoredPasswordAvailableException()))
    monkeypatch.setattr(utils.getpass, "getpass", lambda prompt: "public_secret")
    out = utils.get_password("userx", interactive=True)
    assert out == "public_secret"

File: RepoTransBench/target_projects/Python/Rust/picklepete_pyicloud/public_tests/test_public_utils.rs
use std::collections::HashMap;
use std::cell::RefCell;
use std::rc::Rc;

thread_local! {
    static SAVED: RefCell<HashMap<String, String>> = RefCell::new(HashMap::new());
    static DELETED: RefCell<Vec<String>> = RefCell::new(Vec::new());
}

struct PublicDummyKeyring;

impl PublicDummyKeyring {
    fn get_password(system: &str, username: &str) -> Option<String> {
        SAVED.with(|s| s.borrow().get(username).cloned())
    }
    fn set_password(system: &str, username: &str, password: &str) -> &'static str {
        SAVED.with(|s| s.borrow_mut().insert(username.to_string(), password.to_string()));
        "store"
    }
    fn delete_password(system: &str, username: &str) -> &'static str {
        DELETED.with(|d| d.borrow_mut().push(username.to_string()));
        "removed"
    }
}

fn get_password_from_keyring(username: &str) -> Result<String, &'static str> {
    PublicDummyKeyring::get_password("icloud", username).ok_or("PyiCloudNoStoredPasswordAvailableException")
}

fn password_exists_in_keyring(username: &str) -> bool {
    PublicDummyKeyring::get_password("icloud", username).is_some()
}

fn store_password_in_keyring(username: &str, password: &str) -> &'static str {
    PublicDummyKeyring::set_password("icloud", username, password)
}

fn delete_password_in_keyring(username: &str) -> &'static str {
    PublicDummyKeyring::delete_password("icloud", username)
}

fn underscore_to_camelcase(s: &str, initial_capital: bool) -> String {
    let mut iter = s.split('_');
    let mut result = String::new();
    if initial_capital {
        for word in iter {
            let mut chars = word.chars();
            if let Some(first) = chars.next() {
                result.push_str(&first.to_string().to_uppercase());
                result.push_str(&chars.as_str().to_lowercase());
            }
        }
    } else {
        if let Some(first) = iter.next() {
            result.push_str(&first.to_lowercase());
        }
        for word in iter {
            let mut chars = word.chars();
            if let Some(first) = chars.next() {
                result.push_str(&first.to_string().to_uppercase());
                result.push_str(&chars.as_str().to_lowercase());
            }
        }
    }
    result
}

#[test]
fn test_get_password_from_keyring_success() {
    SAVED.with(|s| s.borrow_mut().insert("baz".to_string(), "qux".to_string()));
    assert_eq!(get_password_from_keyring("baz").unwrap(), "qux".to_string());
}

#[test]
fn test_get_password_from_keyring_failure() {
    let result = get_password_from_keyring("does-not-exist");
    assert!(result.is_err());
}

#[test]
fn test_password_exists_in_keyring_true() {
    SAVED.with(|s| s.borrow_mut().insert("public_user".to_string(), "public_pw".to_string()));
    assert!(password_exists_in_keyring("public_user"));
}

#[test]
fn test_password_exists_in_keyring_false() {
    assert!(!password_exists_in_keyring("anonymous"));
}

#[test]
fn test_store_password_in_keyring() {
    let out = store_password_in_keyring("user2", "passwd2");
    let found = SAVED.with(|s| s.borrow().get("user2").map(|v| v == "passwd2").unwrap_or(false));
    assert!(found, "Password not stored in keyring");
    assert_eq!(out, "store");
}

#[test]
fn test_delete_password_in_keyring() {
    SAVED.with(|s| s.borrow_mut().insert("deleteme".to_string(), "secret".to_string()));
    let out = delete_password_in_keyring("deleteme");
    let deleted = DELETED.with(|d| d.borrow().contains(&"deleteme".to_string()));
    assert!(deleted);
    assert_eq!(out, "removed");
}

#[test]
fn test_underscore_to_camelcase_basic() {
    assert_eq!(underscore_to_camelcase("foo_bar_baz", false), "fooBarBaz");
    assert_eq!(underscore_to_camelcase("Bar_c", true), "BarC");
}

#[test]
fn test_get_password_interactive_false() {
    fn fail_get_password_from_keyring(_username: &str) -> Result<String, &'static str> {
        Err("PyiCloudNoStoredPasswordAvailableException")
    }
    let res = fail_get_password_from_keyring("userx");
    assert!(res.is_err());
}

#[test]
fn test_get_password_interactive_true() {
    fn fail_get_password_from_keyring(_username: &str) -> Result<String, &'static str> {
        Err("PyiCloudNoStoredPasswordAvailableException")
    }
    fn getpass(_prompt: &str) -> String {
        "public_secret".to_string()
    }
    let out = getpass("Prompt: ");
    assert_eq!(out, "public_secret");
}